server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,450 @@
+}
+
+impl PrompterCallback {
+    pub async fn new(window_id: &str, service: Service, prompt_path: OwnedObjectPath) -> Self {
+        let index = service.prompt_index().await;
+        Self {
+            window_id: window_id.to_owned(),
+            secret_exchange_pub_key: Default::default(),
+            secret_exchange_aes_key: Default::default(),
+            path: OwnedObjectPath::try_from(format!("/org/gnome/keyring/Prompt/p{index}")).unwrap(),
+            service,
+            prompt_path,
+        }
+    }
+
+    pub fn path(&self) -> &OwnedObjectPath {
+        &self.path
+    }
+
+    pub async fn object_label(&self, objects: &Vec<OwnedObjectPath>) -> Option<String> {
+        for object in objects {
+            if let Some(collection) = self.service.collection_from_path(object).await {
+                return Some(collection.label().await);
+            }
+        }
+        None
+    }
+
+    pub async fn set_secret_exchange_pub_key(&self, pub_key: String) {
+        *self.secret_exchange_pub_key.write().await = pub_key;
+    }
+
+    pub async fn secret_exchange_pub_key(&self) -> String {
+        self.secret_exchange_pub_key.read().await.clone()
+    }
+
+    pub async fn set_secret_exchange_aes_key(&self, aes_key: String) {
+        *self.secret_exchange_aes_key.write().await = aes_key;
+    }
+
+    pub async fn secret_exchange_aes_key(&self) -> String {
+        self.secret_exchange_aes_key.read().await.clone()
+    }
+
+    pub async fn perform_prompt(
+        connection: zbus::Connection,
+        path: OwnedObjectPath,
+        prompt_type: PromptType,
+        properties: Properties,
+        exchange: String,
+    ) -> Result<(), ServiceError> {
+        let prompter = PrompterProxy::new(&connection).await?;
+        prompter
+            .perform_prompt(path, prompt_type, properties, &exchange)
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn stop_prompting(
+        connection: zbus::Connection,
+        path: OwnedObjectPath,
+    ) -> Result<(), ServiceError> {
+        let prompter = PrompterProxy::new(&connection).await?;
+        prompter.stop_prompting(path).await?;
+
+        Ok(())
+    }
+
+    pub async fn prompt_completed(
+        signal_emitter: zbus::object_server::SignalEmitter<'_>,
+        dismissed: bool,
+        result: OwnedValue,
+    ) -> Result<(), ServiceError> {
+        Prompt::completed(&signal_emitter, dismissed, result).await?;
+        tracing::debug!("Prompt completed.");
+
+        Ok(())
+    }
+
+    pub async fn prompt(
+        &self,
+        prompt_role: &PromptRole,
+        reply: Optional<Reply>,
+        exchange: &str,
+        connection: zbus::Connection,
+        prompt_path: OwnedObjectPath,
+        objects: Vec<OwnedObjectPath>,
+    ) -> Result<(), ServiceError> {
+        let locked = match prompt_role {
+            PromptRole::Lock => true,
+            PromptRole::Unlock => false,
+            PromptRole::CreateCollection => todo!(),
+        };
```

Comment:
I am not sure this is needed. The role is enough.

Reply:
Sorry, I don't get it.

Reply:
I don't get what you wrote here neither. What I said is there is 0 reason to convert an enum that already conveys the information properly to a boolean